lib/puppet/pops/merge_strategy.rb

```diff
@@ -0,0 +1,224 @@
+      e2.merge(e1)
+    end
+
+    def key
+      :hash
+    end
+
+    protected
+
+    def value_t
+      @value_t ||= Puppet::Pops::Types::TypeParser.new.parse('Hash[String,Data]')
+    end
+
+    MergeStrategy.add_strategy(self)
+  end
+
+  # Merges two values that must be either scalar or arrays into a unique set of values.
+  #
+  class UniqueMergeStrategy < MergeStrategy
+    # Merges two values that must be either scalar or arrays into a unique set of values.
+    #
+    # Scalar values will be converted into a one element arrays and array values will be flattened
+    # prior to forming the unique set. The order of the elements is preserved with e1 being the
+    # first contributor of elements and e2 the second.
+    #
+    # @param e1 [Array<Object>] The first array
+    # @param e2 [Array<Object>] The second array
+    # @return [Array<Object>] The unique set of elements
+    #
+    def checked_merge(e1, e2, options)
+      to_flat_a(e1) | to_flat_a(e2)
+    end
+
+    def key
+      :unique
+    end
+
+    protected
+
+    def value_t
+      @value_t ||= Puppet::Pops::Types::TypeParser.new.parse('Variant[Scalar,Array[Data]]')
+    end
+
+    def to_flat_a(e)
+      e.is_a?(Array) ? e.flatten : [e]
+    end
+
+    MergeStrategy.add_strategy(self)
+  end
+
+  begin
+    # Avoid using the full gem since adds methods to the core Hash class
+    require 'deep_merge/core'
+
+    # Performs a deep merge. The arguments can be either Hash or Array
+    #
+    # @see https://github.com/danielsdeleo/deep_merge
+    #
+    class DeepMergeStrategy < MergeStrategy
+      def checked_merge(e1, e2, options)
+        # e1 (the destination) is cloned to avoid that the passed in object mutates
+        DeepMerge.deep_merge!(e1, e2.clone, { :preserve_unmergeables => false }.merge(options))
+      end
+
+      def key
+        :deep
+      end
+
+      protected
+
+      # Returns a type that allows all deep_merge options except 'preserve_unmergeables' since we force
+      # the setting of that option to false
+      #
+      # @return [Puppet::Pops::Types::PAnyType] the puppet type used when validating the options hash
+      def options_t
+        @options_t ||= Puppet::Pops::Types::TypeParser.new.parse('Struct[{'\
+            "strategy=>Optional[Pattern[#{key}]],"\
+            'knockout_prefix=>Optional[String],'\
+            'merge_debug=>Optional[Boolean],'\
+            'merge_hash_arrays=>Optional[Boolean],'\
+            'sort_merge_arrays=>Optional[Boolean],'\
+            'unpack_arrays=>Optional[String]'\
+            '}]')
+      end
+
+      def value_t
+        @value_t ||= Puppet::Pops::Types::TypeParser.new.parse('Variant[Array[Data],Hash[String,Data]]')
+      end
+
+      MergeStrategy.add_strategy(self)
+    end
+  rescue LoadError
+    # Deep merge missing. OK but let's print a debug statement
+    Puppet.debug("Merge strategy for 'deep' merge is not present because the deep_merge gem is not present")
```

Comment:
This should raise an error instead (see comments about exception handling elsewhere how it should be handled).
We expect the deep merge gem to always be there in Puppet 4.0.